### ISSUE TITLE
Inconsistent reading and writing from files because of state names

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#1221]
By default, pgmpy assigns an `int` type state name to variables if no state names are explicitly specified. This creates a problem when these models are written to a file and read back. Since there is no way to differentiate between an `int` type state or `str` type state (i.e whether the state was `0` or `'0'` when written) when reading from the file, it can lead to unexpected results. The reader functions currently always by default assume the state to be of `str` type, so the state `0` when written and read back will be `'0'`.

A possible solution to this is to change the default state name to `str` type instead of `int`.

**Comment by loudly-soft** [comment]
If I understand correctly, even if state names are explicitly specified, reading and writing using Bif  causes the same, if not, similar problem.  Perhaps this is the same issue as #38

For example, the code below throws this error:

> pgmpy/utils/state_name.py", line 74, in get_state_no
>     return self.name_to_no[var][state_name]
> KeyError: 2
> 

```
from pgmpy.models import BayesianModel
from pgmpy.estimators import BayesianEstimator
from pgmpy.readwrite import BIFWriter, BIFReader
import pandas as pd

# define data
df_data = pd.DataFrame([{'A': 1, 'B': 2, 'C': 1}], columns=['A', 'B', 'C'])

# train
model = BayesianModel([('A', 'B'), ('A', 'C')])
model.fit(df_data, state_names={'A':[1, 2], 'B':[1, 2], 'C':[1, 2]}, estimator=BayesianEstimator, prior_type='K2')

# save model
writer = BIFWriter(model)
writer.write_bif("bif_test.bif")

# load model
reader = BIFReader("bif_test.bif")
model = reader.get_model()

# predict
df_test = pd.DataFrame([{'B': 2, 'C': 2}], columns=['B', 'C'])
print(model.predict_probability(df_test))
```

**Comment by ankurankan** [comment]
@loudly-soft There are two problems here, the first is that the state names were not getting saved properly by the BIFWriter which has now been fixed #1255. The second is because of the int state name while saving and str when reading. For the time being, I would suggest using just str state names until the issue is fixed.
